.github/ISSUE_TEMPLATE/feature_request.md

```diff
@@ -1,21 +1,17 @@
 ---
-name: Feature request
-about: Suggest a new idea for Jamulus (not an improvement of an existing function)
+name: Feature
+about: Agreed specificaiton for a feature request or enhancement
 title: ''
-labels: 'feature request'
+labels: 'feature'
```

Comment:
Are you sure? Is this the new label for feature requests?

Reply:
This is what's been being discussed.  Feature requests, i.e. anything that isn't defined and accepted as a new feature that's going to be delivered, get talked about in Discussions.  Then a formal feature issue gets raised with the agreed spec.

Reply:
Yes, looking at the Issues yesterday, I'd say maybe 80% of them were "Hey! Lets do X!" followed by people saying, "No! Why? What do you mean?" then after about 20-30 messages it would be "OK, so about we do Y then?". But meanwhile, the original "Let's do X!" was intact so it was impossible to really know what the "Issue" was, let alone how to label it for a release, tie it to a PR or whatever.

Reply:
I fully agree with this direction, however I'm a bit unsure where to draw the line. If Github issues should be strictly limited to perfectly specified new features, then I guess this would already involve half of the ideas about the specific implementation. This would also mean that issues would practically be more or less limited to people who intend to write the actual code.

This has certainly benefits, but I wonder whether that would be a bit too strict.

There are issues where the intention is absolutely clear, but there is still some discussion/opinions regarding risk of implementation or specific details. A good example is https://github.com/jamulussoftware/jamulus/issues/1016. In that case I think I would have preferred for it to stay an issue as the "discussion" around it is valuable in the context of the implementation?

On the other hand, I fully agree that issues like the [big metronome one](https://github.com/jamulussoftware/jamulus/discussions/1064) should really be discussions until everything is sorted.

I guess the problem is that it will be hard to predict when issues go "out of control". I fear that we might direct too many to discussions for such requests. This would keep the issue list clean, but would make it harder to manage all the open requests which would then be in the discussions area?

Reply:
> If Github issues should be strictly limited to perfectly specified new features,

Sorry, what I meant was that they should be at least reasonably stable. Not literally morphing into different things over time. 

> This would also mean that issues would practically be more or less limited to people who intend to write the actual code

I also didn't mean to imply that people can't just do PRs and ask for forgiveness later :-) It's just that if you use the word "request" it implies discussion, and then it all goes down hill from there. 

So - a rule of thumb perhaps is: does your issue come with some kind of implementation guidance such that somebody who wasn't party to the discussion might pick it up and do a PR for it? A question, as with #1016 like "can it be done with Qt without writing platform-specific code?" is exactly the sort of thing that may lead to "OK let's do Y then instead" (possibly after numerous posts making it very hard to work out what's happening to the ticket).

Reply:
Thinking some more about the original template, I think there's some wisdom in perhaps reminding the user to list alternatives that have been considered, since that would be good to to effectively summarise the issues in the discussion that usually go "Let's do X!" "No, because [reason]" "OK let's do Y then". That might prevent people wanting to discuss it all over again in the Issue.

Reply:
I really like the current version of the PR! :)

> It's just that if you use the word "request" it implies discussion, and then it all goes down hill from there.

Agreed.

> A question, as with #1016 like "can it be done with Qt without writing platform-specific code?" is exactly the sort of thing that may lead to "OK let's do Y then instead" (possibly after numerous posts making it very hard to work out what's happening to the ticket).

I think we are heading in the same direction, but have different tendencies where to draw the line. "Can it be done with Qt" sounds highly implementation-specific to me.
My mental model would be essentially, sorted from vague to specific:
- **Discussions** start with vague ideas, include discussions about different approaches and get lots of (valuable!) side-discussions from other people chiming in. Example "We should have input gain adjustments in Jamulus"
- **Issues** would be the "Intent to implement", e.g. "I want to make Jamulus work for people who have no other means to adjust their input volume. I'm therefore going to allow the server to mix people's input above 100%".
- **PRs**  obviously show a concrete way to implement a given feature. They will still be the place where code-centric discussions will (and IMHO should) happen, e.g. "Why did you implement the gain adjustment in function X instead of Y?". A real-world example for the "should we use Qt or a native function" is #967.

I understand that the goal is to keep issues manageable and actionable in order to make it possible to do some project management on them. The current status is -- or was, thanks to your cleanup work! -- that there is a high number of open issues, some which receive lots of attention, some with almost none. So I fully understand that we should err on the side of keeping things a bit too strict for now.

Anyway, just wanted to share my thoughts and I'm certainly open to try it out as documented in this PR. We could still adjust later.

Reply:
>  "Can it be done with Qt" sounds highly implementation-specific to me.

Jamulus is based on Qt libraries for cross-platform support.  So it's the opposite of being implementation-specific, really.  The _aim_ is to use Qt facilities where available, to _avoid_ writing implementation-specific code.

Reply:
> Jamulus is based on Qt libraries for cross-platform support. So it's the opposite of being implementation-specific, really. The aim is to use Qt facilities where available, to avoid writing implementation-specific code.

I think the aim is clear. I was trying to refer to the stage where this is relevant: Such questions are tied to the actual implementation stage (most likely in the PR itself) and are unlikely to come up in the preceding discussion about a new future. All I'm saying is that pointing someone to use Qt instead of native APIs is not that groundbreaking that it would require an existing issue to be converted "back" to a discussion. In other words, I'd consider this to be a *coding detail* (in your words).

Just trying to clear up the misunderstanding. I really don't want to be the one bike-shedding over such details. :)

Reply:
No, the discussion about the implementation _DOES NOT_ happen in the PR.  The PR is discussion about the specific code submitted.  The _proposed implementation_ should have already been agreed _before_ any code is written - otherwise it's a waste of time and potentially annoying and disappointing for contributors.

Reply:
Yes, I fully agree that this is how it is supposed to be. I was merely describing what I observe in issues/PR currently.

As this PR is about improving exactly this and defining how it should be, I should probably have raised my concerns elsewhere.

Reply:
In @hoffie's defence though, I think we shouldn't really discourage "drive-by PRs" too much in the same way as we shouldn't discourage the posting of un-discussed Issues too much. They are both communications of intent after all. For all I know, it may be potentially annoying and disappointing for contributors of issues or discussions if those get tossed out too.

Reply:
I agreed - this is just so that the templates are clearer, though.

Ideally, there would be a PR template that explained the reviewer expectations of how code should have arrived ready for PR, with pointers to relevant documentation, too.

None of this prevents someone "just doing it".  It will then get handled appropriately.